elements/bulbs-carousel/bulbs-carousel.js

```diff
@@ -0,0 +1,174 @@
+import invariant from 'invariant';
+import { BulbsHTMLElement, registerElement } from 'bulbs-elements/register';
+import { moveChildren } from 'bulbs-elements/util';
+
+import './bulbs-carousel.scss';
+
+export const supportsCalcInTransform = (() => {
+  let div = document.createElement('div');
+  div.style.transform = 'translateX(calc(100% - 10px))';
+  return div.style.transform !== '';
+})();
+
+export default class BulbsCarousel extends BulbsHTMLElement {
+  createdCallback () {
+    invariant(
+      this.slider = this.querySelector('bulbs-carousel-slider'),
+      '<bulbs-carousel> MUST contain a <bulbs-carousel-slider> element.'
+    );
+
+    this.currentIndex = 0;
+
+    this.handleClick = this.handleClick.bind(this);
+    this.addEventListener('click', this.handleClick);
+
+    this.track = document.createElement('bulbs-carousel-track');
+
+    moveChildren(this.slider, this.track);
+
+    this.slider.appendChild(this.track);
+
+    this.previousButtons = this.getElementsByTagName('bulbs-carousel-previous');
+    this.nextButtons = this.getElementsByTagName('bulbs-carousel-next');
+  }
+
+  attachedCallback () {
+    this.pageToCarouselItem(this.getActiveCarouselItem());
+    this.applyState();
+  }
+
+  handleClick (event) {
+    if (event.target.closest('bulbs-carousel-previous')) {
+      this.slideToPrevious();
+      this.applyState();
+    }
+
+    if (event.target.closest('bulbs-carousel-next')) {
+      this.slideToNext();
+      this.applyState();
+    }
+  }
+
+  applyState () {
+    let currentPage = this.getCurrentPage();
+    let onFirstPage = this.isOnfirstPage();
+    let onLastPage = this.isOnLastPage();
+    let translate = -100 * currentPage;
+    let itemMargin = this.getItemMargin();
+    let marginCorrection = itemMargin * currentPage;
+
+    // some IE browsers do not support css calc() in the transform property.
+    // For those browsers one uses multiple translateX() declarations.
+    //      Be sad.
+    if (supportsCalcInTransform) {
+      this.track.style.transform = `translateX(calc(${translate}% - ${marginCorrection}px))`;
+    }
+    else {
+      this.track.style.transform = `translateX(${translate}%) translateX(-${marginCorrection}px)`;
+    }
+
+    function toggleAttribute (collection, attribute, toggle) {
+      Array.prototype.forEach.call(collection, (item) => {
+        toggle ? item.setAttribute(attribute, '') : item.removeAttribute(attribute);
+      });
+    }
+
+    toggleAttribute(this.previousButtons, 'disabled', onFirstPage);
+    toggleAttribute(this.nextButtons, 'disabled', onLastPage);
+  }
+
+  get carouselItems () {
+    return this.track.children;
+  }
+
+  get firstItem () {
+    return this.carouselItems[0];
+  }
+
+  getActiveCarouselItem () {
```

Comment:
I  waffled around on these ones.

I'm a fan of getters. I felt that `get caruselItems` and `get firstItem` were obvious cases for getters because they are merely accessing properties. The methods I've prefixed as `getThing` I wasn't so sure about. They are doing potentially more heavy lifting/DOM access.

I'm on the fence, could be pushed either way.